### PR TITLE
Coding branch

### PR DESCRIPTION
**Pull request by phalado**
Sorry for the requests TSEs, now I think it's alright.

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    i = 0
+    while i < length
+      yield(self[i])
+      i += 1
+    end
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_each_with_index
+    i = 0
+    while i < length
+      yield(i, self[i])
+      i += 1
+    end
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.
* [x] my_each_with_index calls the given block once for each element in self
  - the first argument of my_each should be the item and second argument should be the index. You are doing the opposite.

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_select
+    new_array = []
+    my_each do |x|
+      new_array << x if yield(x)
+    end
+    new_array
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_all?
+    answer = true
+    my_each do |x|
+      answer = false unless yield(x)
+    end
+    answer
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.
* [x] the method should be able to take classes, regex exp, primitive types values as arguments, yours raise an error when do so :red_circle: 

example:
```ruby
    array = [1,2,5,4,7]
    p array.my_all?(Integer)
    # return this error
    #ArgumentError:
    #   wrong number of arguments (given 1, expected 0)
     # ./enumerable.rb:28:in `my_all?'

  ```

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_any?
+    answer = false
+    my_each do |x|
+      answer = true if yield(x)
+    end
+    answer
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.
* [x] the method should be able to take classes, regex exp, primitive types values as arguments, yours raise an error when do so :red_circle: 

example:
```ruby
    array = [1,2,5,4,7]
    p array.my_any?(Integer)
    # return this error
    #ArgumentError:
    #   wrong number of arguments (given 1, expected 0)
     # ./enumerable.rb:36:in `my_any?'

  ```

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_none?
+    answer = true
+    my_each do |x|
+      answer = false if yield(x)
+    end
+    answer
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.
* [x] the method should be able to take classes, regex exp, primitive types values as arguments, yours raise an error when do so :red_circle: 

example:
```ruby
    array = [1,2,5,4,7]
    p array.my_none?(Integer)
    # return this error
    #ArgumentError:
    #   wrong number of arguments (given 1, expected 0)
     # ./enumerable.rb:44:in `my_none?'

  ```

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_count
+    counter = 0
+    my_each do |x|
+      counter += 1 if yield(x)
+    end
+    counter
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] If no block is given (i.e check with block_given? at the beginning of the method) the method should return an Enumerator.
* [x] the method should be able to take literals as arguments, yours raise an error when do so :red_circle: 

example:
```ruby
    array = [1,2,5,4,7]
    p array.my_count # should print the number of items
    # return this error
    #ArgumentError:
    #   wrong number of arguments (given 1, expected 0)
     # ./enumerable.rb:52:in `my_count'

  ```

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      my_each do |x|
+        new_array << yield(x)
+      end
+    else
+      my_each do |x|
+        new_array << param.call(x)
+      end
+    end
+    new_array
+  end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] the method should return an Enumerator in the example below
example:
```ruby
    array = [1,2,5,4,7]
    p array.my_map
    # return this error
    # NoMethodError:
      #  undefined method `call' for nil:NilClass
      # ./enumerable.rb:68:in `block in my_map'
      # ./enumerable.rb:7:in `my_each'
      # ./enumerable.rb:67:in `my_map'
  ```

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    my_each do |x|
+      answer = yield(answer, x)
+    end
+    answer
+  end
+
+  def multiply_els
+    my_inject { |x, y| x * y }
+  end
+end
```
Comment (stratospherique) [inline enumerable.rb]: * [x] the method should be able to take literals, to_proc symbols(e.g :+, :-, etc... ) as arguments, yours raise an error when doing so red_circle

```ruby
range = Range.new(5,50)
range.my_inject(4) { |prod, n| prod * n }
      # return this error
      # ArgumentError:
      #  wrong number of arguments (given 1, expected 0)
      # ./enumerable.rb:74:in `my_inject'
```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Good start @phalado :clap: :clap: However, some requirements needs to be implemented in order for us to approve your project. check below :arrow_down: 

- **[REQUIRED]** Check the code comments below, use the notes to refactor you methods :nerd_face: 

- **[REQUIRED]** Remove all debugging code (e.g commented code, etc...)

### That's it :relaxed: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job @phalado :+1: 
I see you implemented previously suggested requirements. That's good.
However, we can't merge this one yet because some edge cases are still failing. See my comments and please do your best to implement the suggestions to the best of your ability.

### Martin Afani

File: enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    answer = true
+    if block_given?
+      my_each { |x| answer = false unless yield(x) }
+    elsif pat
+      my_each { |x| answer = false unless pattern?(x, pat) }
+    else
+      my_each { |x| answer = false unless x }
+    end
+    answer
+  end
```
Comment (whiz25) [inline enumerable.rb:42]: - [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x]  when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    answer = false
+    if block_given?
+      my_each { |x| answer = true if yield(x) }
+    elsif pat
+      my_each { |x| answer = true if pattern?(x, pat) }
+    else
+      my_each { |x| answer = true unless x }
+    end
+    answer
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    answer = true
+    if block_given?
+      my_each { |x| answer = false if yield(x) }
+    elsif pat
+      my_each { |x| answer = false if pattern?(x, pat) }
+    else
+      my_each { |x| answer = false if x }
+    end
+    answer
+  end
```
Comment (whiz25) [inline enumerable.rb:66]: - [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  def my_inject(*args)
+    answer, sym = inj_param(*args)
+    answer = yield(0, 1).zero? ? 1 : 0
+    if block_given?
+      my_each { |x| answer = yield(answer, x) }
+    elsif arg
+      my_each { |x| answer = answer.public_send(sym, x) }
+    end
+    answer
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [x] combines all elements of enum by applying a binary operation, specified by a block
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by uche-inyama — CHANGES_REQUESTED** [review]
Hi Phalado!

Your methods are still failing the tests pointed out by my colleague, whiz25. 

I would suggest you go through this [document](https://ruby-doc.org/core-2.6.5/Enumerable.html). Use the test-cases in this doc, for similar methods, to test the methods you have written. Once the methods you have written pass the tests you should be fine.

When you are done submit another form for a code review.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Comment by phalado** [comment]
> Hi Phalado!
> 
> Your methods are still failing the tests pointed out by my colleague, whiz25.
> 
> I would suggest you go through this [document](https://ruby-doc.org/core-2.6.5/Enumerable.html). Use the test-cases in this doc, for similar methods, to test the methods you have written. Once the methods you have written pass the tests you should be fine.
> 
> When you are done submit another form for a code review.
> 
> #### `Have a great day` 🙂
> `Uchechukwu Henry Inyama`

**Comment by phalado** [comment]
> Hi Phalado!
> 
> Your methods are still failing the tests pointed out by my colleague, whiz25.
> 
> I would suggest you go through this [document](https://ruby-doc.org/core-2.6.5/Enumerable.html). Use the test-cases in this doc, for similar methods, to test the methods you have written. Once the methods you have written pass the tests you should be fine.
> 
> When you are done submit another form for a code review.
> 
> #### `Have a great day` 🙂
> `Uchechukwu Henry Inyama`

What tests are failing? All the tests that my partner and I have done are succeeding.

**Comment by uche-inyama** [comment]
Hi Phalado!

You can reach me slack for clarifications. @ Inyama, Uchechukwu Henry

**Review by wathigo — APPROVED** [review]
# Approved :smile: 

Hi Devs,
Great work implementing the suggested changes :clap: 

### You may move on :next_track_button: 

_Good luck in your future projects :crossed_fingers: 
Reviewed by: **Simon Wathigo**_

**Pull request merged**